### ISSUE TITLE
TYPE_CHECKING in an assignment expression in if statement makes code unreachable

### ISSUE THREAD

**Issue by roundedrectangle** [microsoft/pyright#10503]
**Describe the bug**
`TYPE_CHECKING` variable can be imported from the `typing` module or defined manually like this:

```python
TYPE_CHECKING = False
```

No matter how it is done, it will be `True` for the type checker when used in statements such as `if`. For example:

```python
TYPE_CHECKING = False
if TYPE_CHECKING:
    ... # code here will never be ran in reality, but type checker will think that it will
```

[An assignment expression](https://peps.python.org/pep-0572/), introduced in Python 3.8, can do same thing but one line less. For example:

```python
if TYPE_CHECKING := False:
    ...
```

While both code examples do the same thing, in reality the second code example becomes marked as unreachable for the type checker, which is unexpected.

**Code or Screenshots**

```python
TYPE_CHECKING = False
if TYPE_CHECKING:
    pass # code is reachable for the type checker, as expected


# code doing same thing but one line less:
if TYPE_CHECKING := False:
    pass # code is unreachable for the type checker
```

**VS Code extension or command-line**
Pyright 1.1.401, Python 3.13

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

The `TYPE_CHECKING` symbol is special in python typing. It is assumed to be constant, and it is handled very early in the type checking process (in the "binding" phase), prior to any type analysis. This is because conditionals that use this symbol are used to build the code flow graph, and the code flow graph is needed for type analysis. Pyright's binder supports a limited set of syntactic forms which are documented [here](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=static-conditional-evaluation). These forms are dictated by the Python typing spec. If you deviate from these, you will not get the desired behavior. 

To make your code type check, you should avoid using an assignment expression to assign a value to `TYPE_CHECKING`. Use the two-line form.
